src/Data/HashMap/Linear.hs

```diff
@@ -59,218 +58,185 @@ import qualified Prelude
 --------------------------------------------------
 
--- | Run a computation using a singleton hashmap
-singleton :: Keyed k =>
-  (k, v) -> (HashMap k v #-> Unrestricted b) -> Unrestricted b
-singleton (k :: k, v :: v) (f :: HashMap k v #-> Unrestricted b) =
-  alloc defaultSize (k, v, -1) applyHM
-  where
-    applyHM :: Array (RobinVal k v) #-> Unrestricted b
-    applyHM arr = let ixToHash = (hash k) `mod` defaultSize in
-      f $ HashMap (defaultSize, 1) (write arr ixToHash (k, v, 0))
-
-singleton' :: Keyed k =>
-  (k, v) -> (HashMap k v #-> Unrestricted b) -> b
-singleton' kv f = unUnrestricted (singleton kv f)
-
--- XXX: re-write linearly
--- | Given a key @k@ to 'lookup', look up a @Maybe v@ and feed it to
--- a function @Maybe v -> Maybe v@, and if the result is @Nothing@,
--- delete that key, and otherwise, replace the value with the @v@
--- in the returned @Just v@.
-alter ::
-  Keyed k =>
-  HashMap k v #-> (Maybe v -> Maybe v) ->
-  k ->
-  HashMap k v
-alter = Unsafe.toLinear unsafeAlter
+-- | Run a computation given an empty hashmap
+empty :: forall k v b.
+  Keyed k => Size -> (HashMap k v #-> Unrestricted b) -> Unrestricted b
+empty sz@(Size s) scopeFn =
+  alloc s (-1, (error "key err val", error "val err val")) robinArrayScopeFn
   where
-    unsafeAlter :: Keyed k =>
-      HashMap k v -> (Maybe v -> Maybe v) -> k -> HashMap k v
-    unsafeAlter hmap f k =
-      case lookup hmap k of
-        (hmap', Unrestricted maybeV) -> case f maybeV of
-          Nothing -> delete hmap' k
-          Just v -> insert hmap' k v
-
--- | If the key is present, this update the value,
--- otherwise insert a new key-value pair.
-insert :: Keyed k => HashMap k v #-> k -> v -> HashMap k v
-insert hmap k v = insertFromQuery v $ queryIndex (maybeResize hmap) k
-  where
-    insertFromQuery :: Keyed k =>
-      v -> (HashMap k v, RobinQuery k) #-> HashMap k v
-    insertFromQuery v (HashMap (!size, !count) arr, IndexToInsert (k, psl) ix) =
-      HashMap (size, count + 1) (write arr ix (k, v, psl))
-    insertFromQuery v (HashMap sizes arr, IndexToUpdate (k, psl) ix) =
-      HashMap sizes (write arr ix (k, v, psl))
-    insertFromQuery v (HashMap sizes arr, IndexToSwap (k, psl) ix) =
-      (Unsafe.toLinear swapFromRead) (read arr ix) sizes ix (k, v, psl)
-    -- XXX: This re-does the probe sequence of (k',v')
-    swapFromRead ::
-      Keyed k =>
-      (RobinArr k v, RobinVal k v) ->
-      (Int, Int) ->
-      Int ->
-      RobinVal k v ->
-      HashMap k v
-    swapFromRead (arr', (k', v', _)) sizes ix (k, v, psl) =
-      insert (HashMap sizes (write arr' ix (k, v, psl))) k' v'
-
--- | This deletes the key-value pair if it is present, and otherwise does
--- nothing.
+    robinArrayScopeFn :: RobinArr k v #-> Unrestricted b
+    robinArrayScopeFn arr = scopeFn $ HashMap sz 0 arr
+
+-- | If the key is present, this update the value.
+-- If not, if there's enough space, insert a new pair.
+-- If there isn't enough space, error.
```

Comment:
I'm confused. Why are we erroring when the hashmap is full? Instead of resizing?